doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,984 @@
+    return _analyze(text, analysis_kind)
+```
+
+We can annotate this function with overloads to help inform the type checker (and intellisense) which input maps to
+which output type.
+
+```python
+from typing import overload
+
+
+@overload
+def analyze_text(text: str, analysis_kind: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: SentimentAnalysis) -> SentimentResult:
+    ...
+
+
+def analyze_text(text, analysis_kind):  # the actual implementation shouldn't be annotated
+    return _analyze(text, analysis_kind)
+```
+
+Running mypy shows that the correct return type is inferred from the input type:
+
+```python
+result = analyze_text("my text", LanguageDetection())
+reveal_type(result)
+```
+
+`main.py:28: note: Revealed type is "__main__.LanguageDetectionResult"`
+
+Another use of typing.overload is when there are a different number of arguments. We can take the above example and
+imagine that `analyze_text` has an additional overload which does not take an `analysis_kind`:
+
+```python
+from typing import overload
+
+
+@overload
+def analyze_text(text: str, analysis_kind: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: SentimentAnalysis) -> SentimentResult:
+    ...
+
+
+@overload
+def analyze_text(text: str) -> str:
+    ...
+
+
+def analyze_text(*args, **kwargs):  # the actual implementation shouldn't be annotated
+    # logic parsing args / kwargs
+    ...
+```
+
+Running mypy again shows that it understands the return types based on the different number of arguments:
+
+```python
+language_result = analyze_text("my text", LanguageDetection())
+reveal_type(language_result)
+
+text_result = analyze_text("my text")
+reveal_type(text_result)
+```
+
+```cmd
+main.py:36: note: Revealed type is "__main__.LanguageDetectionResult"
+main.py:39: note: Revealed type is "builtins.str"
+```
+
+### Use typing.cast to help mypy understand a type
+
+Sometimes `mypy` will be unable to infer a type and may raise a false positive error. Using `typing.cast` is a way to
+tell `mypy` what the type of something is and is generally favored over using a `# type: ignore`.
+
+The below snippet shows an example of how `mypy` might need a little help to understand the parameter type specified in
+a union:
+
+```python
+from typing import Union, List, Optional, Any, cast
+
+
+def receive_deferred_messages(
+        self, sequence_numbers: Union[int, List[int]], *, timeout: Optional[float] = None, **kwargs: Any
+) -> List[ServiceBusReceivedMessage]:
+    if isinstance(sequence_numbers, int):
+        sequence_numbers = [sequence_numbers]
+    sequence_numbers = cast(List[int], sequence_numbers)  # mypy clarity
```

Comment:
for clarity that this is being casted b/c list operations won't be valid on an "Int" type, can we include a list operation line after this? something like?
```suggestion
    sequence_numbers = cast(List[int], sequence_numbers)  # mypy clarity
    
    for num in sequence_numbers:    # if not casted as List[int], mypy error
```

Reply:
sure!